opa-audit-operator/tests/integration/test_charm.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from pathlib import Path
+import os
+import pytest
+import yaml
+from kubernetes import client, config
+from kubernetes.client.rest import ApiException
+from jinja2 import Template
+
+log = logging.getLogger(__name__)
+
+meta = yaml.safe_load(Path("metadata.yaml").read_text())
+
+KUBECONFIG = os.environ.get("TESTING_KUBECONFIG", "~/.kube/config")
+
+config.load_kube_config(KUBECONFIG)  # TODO: check how to use this with designated file
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test):
+    charm = await ops_test.build_charm(".")
+
+    role_binding_file = Path("/tmp/k8s-rolebinding.yaml")
+
+    model_name = ops_test._default_model_name
+
+    with open("docs/gatekeeper-rb.yaml.template", "r") as fh:
+        template = Template(fh.read())
+        role_binding_file.write_text(
+            template.render(
+                service_account_user=f"system:serviceaccount:{model_name}:kubernetes-operator"
+            )
+        )
+    role_binding = yaml.load_all(role_binding_file.read_text(), Loader=yaml.FullLoader)
+    with client.ApiClient() as api_client:
+        api_instance = client.RbacAuthorizationV1Api(api_client)
+        try:
+            for k8s_obj in role_binding:
+                if k8s_obj["kind"] == "ClusterRoleBinding":
+                    api_instance.create_cluster_role_binding(body=k8s_obj)
+                if k8s_obj["kind"] == "ClusterRole":
+                    api_instance.create_cluster_role(body=k8s_obj)
+        except ApiException as err:
+            if err.status == 409:
+                # ignore "already exists" errors so that we can recover from
+                # partially failed setups
+                return
+            else:
+                raise
+
+    ca_cert = ""
+    with client.ApiClient() as api_client:
+        # Create an instance of the API class
+        api_instance = client.CoreV1Api(api_client)
+        name = f"{model_name}-gatekeeper-validating-webhook-configuration"
+
+        ca_cert = api_instance.read_namespaced_secret(name, model_name)
```

Comment:
This call fails with a 404 response:

```
____________________________________________ test_build_and_deploy _____________________________________________
Traceback (most recent call last):
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/tests/integration/test_charm.py", line 57, in test_build_and_deploy
    ca_cert = api_instance.read_namespaced_secret(name, model_name)
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/api/core_v1_api.py", line 24017, in read_namespaced_secret
    return self.read_namespaced_secret_with_http_info(name, namespace, **kwargs)  # noqa: E501
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/api/core_v1_api.py", line 24112, in read_namespaced_secret_with_http_info
    return self.api_client.call_api(
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/api_client.py", line 348, in call_api
    return self.__call_api(resource_path, method,
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/api_client.py", line 180, in __call_api
    response_data = self.request(
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/api_client.py", line 373, in request
    return self.rest_client.GET(url,
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/rest.py", line 239, in GET
    return self.request("GET", url,
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-audit-operator/.tox/integration/lib/python3.8/site-packages/kubernetes/client/rest.py", line 233, in request
    raise ApiException(http_resp=r)
kubernetes.client.exceptions.ApiException: (404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '22fddc52-adb9-40f8-a89c-0e3acbf094bd', 'X-Kubernetes-Pf-Prioritylevel-Uid':
'cde87f0a-f37d-49ee-8701-e6224c371cc0', 'Date': 'Tue, 20 Jul 2021 17:38:40 GMT', 'Content-Length': '292'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"secrets \"test-charm-jox8-gatekeeper-validating-webhook-configuration\" not found","reason":"NotFound","details":{"name":"test-charm-jox8-gatekeeper-validating-webhook-configuration","kind":"secrets"},"code":404}
```

Looks like it's trying to read a secret that doesn't exist yet. Where is this secret supposed to be created?

Reply:
updated